### ISSUE TITLE
Feature request: rescale transformation

### ISSUE THREAD

**Issue by teunbrand** [r-lib/scales#415]
I'd like to propose a new transformation function that essentially wraps the `rescale()` function. It's definition can be something like the following:

``` r
library(scales)

transform_rescale <- function(to = c(0, 1), from = c(0, 1)) {
  force(to)
  force(from)
  new_transform(
    "rescaling",
    transform = function(x) rescale(x, to = to, from = from),
    inverse   = function(x) rescale(x, to = from, from = to) 
  )
}
```
It should be able to handle a bunch of 'scale and translate' transformations, like unit conversions. Going from metres to inches is simply a scale transformation.
```r
metres_to_inches <- transform_rescale(to = c(0, 100/2.54))
metres_to_inches$transform(1)
#> [1] 39.37008
```
Whereas going from degrees celcius to degrees fahrenheit is a 'scale and translate' transformation.
```r
celsius_to_fahrenheit <- transform_rescale(to = c(32, 212), from = c(0, 100))
celsius_to_fahrenheit$transform(-5)
#> [1] 23
celsius_to_fahrenheit$transform(32)
#> [1] 89.6
```

<sup>Created on 2023-12-15 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Given that it is suitable for plenty of unit conversions, I think this transformation may have a bunch of applications.

**Comment by mjskay** [comment]
In case it is desired, adding derivative support for this should entail something like:

```r
transform_rescale <- function(to = c(0, 1), from = c(0, 1)) {
  force(to)
  force(from)
  new_transform(
    "rescaling",
    transform = function(x) rescale(x, to = to, from = from),
    inverse   = function(x) rescale(x, to = from, from = to),
    d_transform = function(x) rep(diff(to) / diff(from), length(x)),
    d_inverse = function(x) rep(diff(from) / diff(to), length(x))
  )
}
```

**Comment by teunbrand** [comment]
Ah yes, of course! Thanks for weighing in, it was a little oversight on my part

**Comment by thomasp85** [comment]
Happy to accept a PR for this :-)

**Comment by teunbrand** [comment]
There was something fundamentally flawed with this approach when I last tried it. Unfortunately, I have forgotten what it was, so let me try again :)

**Comment by teunbrand** [comment]
I've rediscovered why this approach is flawed. Essentially, you calculate breaks and labels in 'original' space, so the transformation is essentially a fancyful `transform_identity()` in disguise.

``` r
library(scales)

transform_rescale <- function(to = c(0, 1), from = c(0, 1)) {
  scales:::force_all(to, from)
  new_transform(
    "rescaling",
    transform   = function(x) rescale(x, to = to, from = from),
    inverse     = function(x) rescale(x, to = from, from = to),
    d_transform = function(x) rep(diff(to) / diff(from), length(x)),
    d_inverse   = function(x) rep(diff(from) / diff(to), length(x))
  )
}

celcius_to_fahrenheit <- transform_rescale(to = c(32, 212), from = c(0, 100))

library(ggplot2)

ggplot(pressure, aes(temperature, pressure)) +
  geom_line() +
  scale_x_continuous(
    transform = celcius_to_fahrenheit
  )
```

![](https://i.imgur.com/6g6SsMc.png)

<sup>Created on 2024-10-22 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Might be easy enough to calculate breaks/labels in transformed space though, then it can have its intended use case.
